jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+                        f"the process with PID: {proc.pid} becomes a zombie and should be considered a dead."
+                    )
+                    return True
+                # the process may be in an uninterruptible sleep
+                j.logger.debug("the process may be in an uninterruptible sleep.")
+                j.logger.warning(f"Could not kill the process with pid: {proc.pid} with {sig}. Timeout: {timeout}")
+                raise j.exceptions.Runtime(f"Could not kill process with pid {proc.pid}, {proc.status()}") from e
+            return True  # XXX only to pass current tests
+        else:
+            j.logger.warning(f"Could not kill the process with pid: {proc.pid} with {sig}. Timeout: {timeout}")
+            raise j.exceptions.Runtime(f"Could not kill process with pid {proc.pid}") from e
+    except psutil.AccessDenied as e:
+        # permission to perform an action is denied
+        j.logger.warning("the permission is denied")
+        raise j.exceptions.Permission("Permission to perform this action is denied!") from e
+    except (psutil.ZombieProcess, psutil.NoSuchProcess):
+        # Process no longer exists or Zombie (already dead)
+        j.logger.debug("Process is no longer exists or a Zombie (already dead)")
         return True
-    except Exception as e:
-        raise j.exceptions.RuntimeError("Could not kill process with id %s.\n%s" % (pid, e))
 
 
-def ps_find(name):
-    """find process by name
+def ps_find(process_name):
+    """Check if there is any running process that match the given name.
 
-    Arguments:
-        name {str} -- process name
+    Args:
+        process_name (str): The target process name. will match against against Process.name(), Process.exe() and Process.cmdline()
 
     Returns:
-        [bool] -- True if process is found
+        bool: True if process is found, False otherwise.
     """
-    for proc in psutil.process_iter():
-        if proc.name() == name:
-            return True
-    return False
+    return len(get_pids(process_name, limit=1)) == 1
 
 
-def kill_all(name, sig=signal.SIGKILL):
-    """Kill all processes with a given name
+def kill_all(process_name, sig=signal.SIGKILL):
+    """Kill all processes that match 'process_name'.
 
-    Arguments:
-        name {str} -- process name
+    use kill_process_by_name is preferred.
+
+    Args:
+        process_name (str): The target process name
+        sig (signal, optional): See signal module constants. Defaults to signal.SIGKILL
 
-    Keyword Arguments:
-        sig (int) -- signal number (default: {signal.SIGKILL})
+    Returns:
+        None or list of int: a list represents the IDs of the processes remaning alive, otherwise None.
     """
-    sig = int(sig)
-    for proc in psutil.process_iter():
-        if proc.name() == name:
-            kill(proc.pid, sig)
+    # XXX kill default to SIGTERM while kill_all default to SIGKILL (inconsistency)?
+    # XXX almost like kill_process_by_name
```

Comment:
If it's the same as `kill_process_by_name` we can remove it and update the references where it's used.

Reply:
so we agree on removing kill_all, right?